src/server.js

```diff
@@ -23,12 +26,86 @@ const sendUserError = (err, res) => {
   }
 };
 
+// const validateUserPassword = server.use((req, res, next) => {
+//   const username = req.body.username;
+//   const { password } = req.body;
+//   console.log(req.body);
+//   if (!username || !password) {
+//     res.status(STATUS_USER_ERROR);
+//     res.json({ error: 'Provide both username and password' });
+//     return;
+//   }
+//   next();
+// });
+
 // TODO: implement routes
+server.post('/users', (req, res) => {
+  const { username, password } = req.body;
+  // checking for a password
+  if(!password || !username) {
+    sendUserError('valid username/password required', res);
+    return;
+  }
+  bcrypt.hash(password, BCRYPT_COST, (err, passwordHash) => {
+    if (err) {
+      sendUserError(err, res);
+      return;
+    }
+    const newuser = new User({ username, passwordHash });
+    newuser.save((err, user) => {
+      if (err) {
+        sendUserError(err, res);
+        return;
+      }
+      res.json(user);
+    });
+  });
+});
+
+server.post('/log-in', (req, res) => {
+  const { username, password } = req.body;
+  const session = req.session;
+  if (!username || !password) {
+    sendUserError('valid username/password required', res);
+  }
+  User.findOne({ username }, (err, user) => {
+    if (err) {
+      sendUserError('no username with that name found', res);
+      return;
+    }
+    if(!user) {
+      sendUserError('username required', res);
+      return;
+    }
+    if(bcrypt.compareSync(password, user.passwordHash)) {
+      session.login = user.username;
+      res.json({ success: true });
+      return;
+    } else {
+    sendUserError('password invalid', res);
+  }
+  });
+});
 
 // TODO: add local middleware to this route to ensure the user is logged in
-server.get('/me', (req, res) => {
+const localMiddleware = (req, res, next) => {
+  const session = req.session;
+  if (!session.login) {
+    sendUserError('Login required', res);
+    return;
+  }
+  User.find({ username: session.login }, (err, user) => {
```

Comment:
Here, I would use `User.findOne` or `User.findById` instead of just `find`. This works fine, but I think it's good to use the slightly more explicit method.